### PR TITLE
Generic function taking closure with inout parameter returns erroneous results

### PR DESCRIPTION
**Pull request by xedin**
Resolves [SR-3520](https://bugs.swift.org/browse/SR-3520).
Resolves [SR-1976](https://bugs.swift.org/browse/SR-1976).
Resolves [SR-3073](https://bugs.swift.org/browse/SR-3073).
Resolves [SR-3479](https://bugs.swift.org/browse/SR-3479).

### CODE REVIEW

**Comment by xedin** [comment]
/cc @rudkx @slavapestov @jtbandes

This is still work in process, but solves crasher related to the ticket and improves diagnostics related to result types of closures with anonymous parameters...

 I'm going to check why doesn't following type-check correctly right now (I'm pretty sure it has to do with how constraint graph contraction works related to parameter binding):

```swift
func sr3520_2<T>(_ item: T, _ update: (inout T) -> Void) {
  var x = item
  update(&x)
}
var i = 42
sr3520_2(i) { $0 += 3 }
```

**Comment by xedin** [comment]
I've figured out what the problem is last night but i don't know yet how to solve it properly - the reason why it's failing to type-check is because `update:` parameter is `(inout <tv_0>)` (where <tv_0> must be materializable) and anonymous closure is going to allocate argument as just `(<tv_1>)` (which might be inout or might not be depending on the solver path) and according to parameter simplification rules following constraint is going to be created to relating parameter and argument `(inout <tv_0>) arg conv (tv_1>)` (keep in mind that tv_0 must be materializable because it's open generic) so now `CG.optimize()` is going to kick-in and merge equivalence classes of `tv_0` and `tv_1` which yields constraint system unsolvable when `tv_1` actually has to be `inout`, because tv_0 must be materializable, but works in other cases e.g.

```swift
func f0<T, U>(_ t: T, _ f: (inout T) -> U) -> U {
  var t2 = t
  return f(&t2)
}

struct X2 {
  func g() -> Float { return 0 }
}

_ = f0(X2(), {$0.g()})
```

since first argument of closure doesn't have to be `inout` in this case, everything works perfectly fine. There is also rule in `CG.optimize` which prohibits such contraction if there is BindParam that binds to strict inout is involved but that's never true in either case.

I've tried to check materializability requirements on both sides and skip contraction if they are different but that breaks a bunch of other cases...

**Comment by slavapestov** [comment]
@xedin We were discussing this problem yesterday because it also came up with https://github.com/apple/swift/pull/6608. @DougGregor and I had the idea that there should be one type variable for the internal type of each parameter (which is an lvalue type for inouts, an array type for varargs, etc) -- and another type variable for the external type of the parameter list (which need not be materializable). A constraint binds the outer type variable with a tuple type of inner type variables, enforcing the necessary conversions. What do you think?

**Comment by xedin** [comment]
@slavapestov That's what I was kind of thinking as well (but less general)...

I thought that maybe if one side of the contractable constraint must be materializable but the other isn't instead of contracting constraint it would make more sense to create a separate binding which is going to reverse the relationship e.g. -> `(inout t0) arg conv (t1)` becomes `t1 bind (inout t0)` this way instead of making type variables equivalent (which _I think_ is wrong) it would establish other type of dependency relationship which is more appropriate.

All that said, I'm not yet sure if that's going to work at all, since I haven't tried it but in theory it should enable t1 to be bound properly and doesn't prevent related 'bind param' constraints from being contracted...

**Comment by jtbandes** [comment]
Perhaps @cwillmor would have thoughts on this; although no longer a maintainer I think he was the original author of BindParam constraint stuff...

**Comment by xedin** [comment]
@slavapestov @DougGregor I've pushed a fix which deals with `inout T0 subtype T1` constraint contraction by `ConstraintGraph.optimize` which breaks system solving because it also contracts bind param bindings which are not supposed to be inout but rather lvalue. The change from subtype to fixed binding made early allows for correct 'bind param' contraction/solving because left-hand side of the constraint is going to be currently marked as 'inout' which marks right-hand side as 'lvalue'. 

Please take a look and let me know what you think, I've added comments and tests for all related issues SR-3520, SR-1976, SR-3073, SR-3479.

**Comment by slavapestov** [comment]
Awesome. So the vararg problem is not related after all?

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by xedin** [comment]
Yeah, vararg is unrelated, it has to do with contraction of subtype constraints, generated by sub-declarations of trailing closures to be precise :)

File: lib/Sema/CSSimplify.cpp [hunk 1]
```diff
@@ -1448,6 +1448,39 @@ ConstraintSystem::matchTypes(Type type1, Type type2, ConstraintKind kind,
       SWIFT_FALLTHROUGH;
 
     case ConstraintKind::Subtype:
+      // Subtype constraints are subject for edge contraction,
+      // which is inappropriate in this case, because it's going to
+      // erase/lose 'inout' modifier after merging equivalence classes
+      // (if inout containts type var, see ConstraintGraph::contractEdges()),
+      // since right-hand side type variable must not be materializable
+      // it can simply get left-hand side as a fixed binding, otherwise fail.
+      if (type1->is<InOutType>() && type1->hasTypeVariable() && typeVar2) {
```
Comment (slavapestov) [inline lib/Sema/CSSimplify.cpp]: Maybe this should be type1->getInOutObjectType()->isTypeVariableOrMember()?

Comment (xedin) [inline lib/Sema/CSSimplify.cpp]: Sure, let me change it to check for type variable directly instead of presence of one.

**Comment by xedin** [comment]
@slavapestov Force pushed the change for type variable checking, thanks!

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test Linux

**Comment by slavapestov** [comment]
@swift-ci Please smoke test Linux

**Comment by xedin** [comment]
@slavapestov previous Linux build is still going, just got to run tests, I took a long time to compile llvm/clang and other pieces...

**Comment by slavapestov** [comment]
@swift-ci Please smoke test Linux

**Comment by xedin** [comment]
\o/

**Comment by xedin** [comment]
@slavapestov Thanks again!

**Comment by xedin** [comment]
@slavapestov Will you resolve the tickets or should I?

**Comment by rudkx** [comment]
@xedin I went ahead and resolved them, thanks!

**Comment by xedin** [comment]
@rudkx Awesome, thank you!

**Comment by timvermeulen** [comment]
This didn't make Swift 3.1?

**Pull request merged**